viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -579,77 +579,58 @@ void SLAMServiceImpl::GetLatestSampledPointCloudMapString(
+
+    // Iterate over image data and add to pointcloud buffer
     int num_points = 0;
-
-    // Sample the image based on the number of pixels. Output is the number of
-    // pixels to skip. skip_count will reduce the size of the PCD to under 32
-    // MB, with additional tuning provided by the samplingFactor. If the PCD
-    //  would already be smaller than 32MB, do not sample the image.
-    // When moving to streaming this behavior may change.
-    int skip_count = (size_data * pixelBytetoPCDByte) / maximumGRPCByteLimit *
-                     samplingFactor;
-    if (skip_count == 0) {
-        skip_count = 1;
-    }
-
     std::string data_buffer;
-
-    // Loop to sample data and reduce resolution. Increments multiplied
-    // by 4 to represent 4 bytes per pixel
-    for (int i = 0; i < size_data; i += skip_count * 4) {
-        // skip pixels that are not in our map(black/past walls)
-        // this check represents [102,102,102]
-        if ((data_vect[i + 0] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 1] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 2] == defaultCairosEmptyPaintedSlice))
-            continue;
-
-        // Determine probability based on color pixel
-        int prob = viam::ViamColorToProbability((int)data_vect[i + 2]);
-        if (prob == 0) continue;
-
-        num_points++;
-
-        int pixel_index = i / 4;
-        int pixel_x = pixel_index % width;
-        int pixel_y = pixel_index / width;
-
-        // Convert pixel location to pointcloud point in meters
-        float x_pos = (pixel_x - painted_slices->origin.x()) * kPixelSize;
-        // Y is inverted to match output from getPosition()
-        float y_pos = -(pixel_y - painted_slices->origin.y()) * kPixelSize;
-        // 2D SLAM so Z is set to 0
-        float z_pos = 0;
-
-        // Turn the map point into a vector to perform transformations with.
-        // Current transformation rotates coordinates to match slam service
-        // expectation (XZ plane)
-        Eigen::Vector3d map_point(x_pos, y_pos, z_pos);
-        auto rotated_map_point = pcdRotation * map_point;
-
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.x());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.y());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.z());
-        viam::utils::writeIntToBufferInBytes(data_buffer, prob);
+    for (int pixel_y = 0; pixel_y < height; pixel_y++) {
+        for (int pixel_x = 0; pixel_x < width; pixel_x++) {
+            // Get bytes associated with pixel
+            int pixel_index = pixel_x + pixel_y * width;
+            int byte_index = pixel_index * bytesPerPixel;
+            std::vector<unsigned char> pixel_data = {
```

Comment:
I might be reading this wrong but it looks like `pixel_data` is always a vector with 2 elements, but are accessing 3 elements in the vector: https://github.com/viamrobotics/viam-cartographer/pull/25/files#diff-92ff8a5cd3a57430d78c6ba3b61a91d1810399b77613eece3c55841868d2822eR644. I don't understand how this can work. What am I missing?

Reply:
Also, I don't understand why this needs to be a vector at all. It appears as if we are only interested in the probability information, which should be a single `unsigned char` right?

Reply:
Great catch! This is using curly bracket initialization, which means it's creating a vector of 2 elements. The program isn't terminating when accessing the 3rd element because std::vector's operator[] is not bound-checked: https://cplusplus.com/reference/vector/vector/operator[]/.

I think we might have meant to use parentheses, like we did before:

```
std::vector<unsigned char> data_vect(data, data + size_data);
```

and as described in example 4 here: https://www.geeksforgeeks.org/initialize-a-vector-in-cpp-different-ways/.

Reply:
I'm happy to change this to a more explicit format however, the code as it stands is doing as intended. When I print out the size and content of pixel_data I get:

`Size of pixel_data: 4 | [102 102 102 204 ]`

Now, this is a bit confusing because this appears to be in the RGBA format not ARGB. But that is a separate issue which I'm still investigating

Reply:
Investigation complete. We are indeed using the ARGB format type, however, due to the big-endian encoding the bytes are arranged as BGRA, meaning we have been using the red channel this entire time for the probability info. The reason we never saw red in the images was due to the JPEG to a string converter setting the pixel in the image to RGB24, see [viam-cartographer/viam-cartographer/src/io/cairo_jpg.cc](https://github.com/viamrobotics/viam-cartographer/blob/main/viam-cartographer/src/io/cairo_jpg.cc)

Reply:
I have added a color struct to reduce confusion in the future.